Quality of Life Improvements

carlkidcrypto opened the pull request:
This PR adds:
* Support for future versions like 20. A PR upstream has been opened. See https://github.com/muttleyxd/clang-tools-static-binaries/pull/56
* Future proofs this repo by automating the downloading of clang-format binary files.

Key Changes:

* Added `download_clang_format_binaries.sh` which automated the downloading of binary files
* Added `maintain.md` to document how the files that  `download_clang_format_binaries.sh` relies on must be maintained

Testing:

I did basic testing of the  `download_clang_format_binaries.sh` script. I tried the following options:

 `./download_clang_format_binaries.sh 19` 
 `./download_clang_format_binaries.sh 18`
 `./download_clang_format_binaries.sh 10`  

I wasn't sure how to test the `action.yml`.

More Thoughts:

This is a big change in terms of how things were done in the past. Perhaps this can be a new major version.
`v1.0.0`. But with a few beta releases first like `v1.0.0b1` or something like that.

Commit:
Automate

- made a shell script: download_clang_format_binaries to grab the bins from the upstream repo. No need to manually commit then all the time.

- added a versions.txt file that maps short names to latest long name. For example 18==18.1.8.

- delete the bins added to this repo. It makes for a large repo to clown down. Plus this should in theory speed up execution time of the action.

- make a blank.txt inside the clang-format folder.

Commit:
Robust

- making the script more robust and future proof.

- rename a few files to add more features to the download script.

Commit:
Documentation

- added a maintain markdown file.

DoozyX commented:
Thanks, the main reason the binaries were kept in the repo was to improve the speed of action, did you test the difference? Anyway going to test the script to improve the current process.

echoix commented on README.md:
````diff
@@ -117,4 +117,4 @@ docker run -it --rm --workdir /src -v $(pwd):/src clang-format-lint \
 ```
 
-Actions clion format binary are from https://github.com/muttleyxd/clang-tools-static-binaries
+Actions cli on format binary are from https://github.com/muttleyxd/clang-tools-static-binaries
````
```suggestion
This action's clang-format binaries are from https://github.com/muttleyxd/clang-tools-static-binaries
```

echoix commented on clang-format/blank.txt:
```diff
@@ -0,0 +1 @@
+This folder is intentionally left blank.
```
Usually, for this pattern, we usually see an empty file named `.gitkeep`, like explained in https://gist.github.com/jlollis/54d3b3a7ed12776ac9f0d4bc13e63ed9

echoix reviewed (commented):
Also, if we are to download files at runtime, we need to be able to verify that the files are exactly the ones intended. For example, before, if someone would want to pin the action to a specific hash, it would correspond exactly to theses binaries. After, there's missing a digest check, and we are then dependent on the availability of another project.

Is there a way to keep the binaries cached (if it's not too big), but also have a way to give it an executable, so in cases where the action is late to publish a new version, we could supply an updated one outside of the action?

carlkidcrypto commented:
> Thanks, the main reason the binaries were kept in the repo was to improve the speed of action, did you test the difference? Anyway going to test the script to improve the current process.

No, I didn't test the speed. I imagine it should be better since the repo to clone down isn't as large anymore 🤔

carlkidcrypto commented:
> Also, if we are to download files at runtime, we need to be able to verify that the files are exactly the ones intended. For example, before, if someone would want to pin the action to a specific hash, it would correspond exactly to theses binaries. After, there's missing a digest check, and we are then dependent on the availability of another project.
> 
> Is there a way to keep the binaries cached (if it's not too big), but also have a way to give it an executable, so in cases where the action is late to publish a new version, we could supply an updated one outside of the action?

I don't think there's a way to keep the binaries cached. We are grabbing them from a different repo.
A possible workaround would be for you to maintain a fork of the repo this one relies on. The fork can be manually updated to the remote repository or manually updated to add missing binaries/desired binaries.

Commit:
Merge branch 'master' into feature/add_20_and_optimize

echoix commented:
> > Thanks, the main reason the binaries were kept in the repo was to improve the speed of action, did you test the difference? Anyway going to test the script to improve the current process.
> 
> 
> 
> No, I didn't test the speed. I imagine it should be better since the repo to clone down isn't as large anymore 🤔 

Note that the whole download is made by the runner in the setup step, and it currently runs inside the same second -- download, extract, copy and all. Using debug logging, we can see the url that is used for a single tarball download, and it is 26.7MB.
I don't feel it is a bottle neck.


```

Download action repository 'DoozyX/clang-format-lint-action@c71d0bf4e21876ebec3e5647491186f8797fde31' (SHA:c71d0bf4e21876ebec3e5647491186f8797fde31)
##[debug]Download 'https://api.github.com/repos/DoozyX/clang-format-lint-action/tarball/c71d0bf4e21876ebec3e5647491186f8797fde31' to '/home/runner/work/_actions/_temp_f0b37bb0-d8ec-4eaf-b629-48a88581b7fc/a71aa249-f4ce-44d5-8b95-56aacbbe0ba8.tar.gz'
##[debug]Unwrap 'DoozyX-clang-format-lint-action-c71d0bf' to '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31'
##[debug]Archive '/home/runner/work/_actions/_temp_f0b37bb0-d8ec-4eaf-b629-48a88581b7fc/a71aa249-f4ce-44d5-8b95-56aacbbe0ba8.tar.gz' has been unzipped into '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31'.
##[debug]action.yml for action: '/home/runner/work/_actions/actions/checkout/11bd71901bbe5b1630ceea73d27597364c9af683/action.yml'.
##[debug]action.yml for action: '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31/action.yml'.
##[debug]Dockerfile for action: '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31/Dockerfile'.
##[debug]Set step '__actions_checkout' display name to: 'Run actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683'
##[debug]Set step '__DoozyX_clang-format-lint-action' display name to: 'Run DoozyX/clang-format-lint-action@c71d0bf4e21876ebec3e5647491186f8797fde31'
##[debug]Set step 'diff' display name to: 'Create and uploads code suggestions to apply'
Complete job name: Formatting Check
```

Compared to the 3 seconds taken to build the dockerfile, and the 15 seconds it takes to run on the repo I use it for (that was way too long with another clang-format action)

Commit:
Address MAIL

- added a gitkeep file
- added more details to the curl command for performance tracking
- updated the target master commit from the upstream repo.
- updated readme

carlkidcrypto commented:
> > > Thanks, the main reason the binaries were kept in the repo was to improve the speed of action, did you test the difference? Anyway going to test the script to improve the current process.
> > 
> > 
> > No, I didn't test the speed. I imagine it should be better since the repo to clone down isn't as large anymore 🤔
> 
> Note that the whole download is made by the runner in the setup step, and it currently runs inside the same second -- download, extract, copy and all. Using debug logging, we can see the url that is used for a single tarball download, and it is 26.7MB. I don't feel it is a bottle neck.
> 
> ```
> 
> Download action repository 'DoozyX/clang-format-lint-action@c71d0bf4e21876ebec3e5647491186f8797fde31' (SHA:c71d0bf4e21876ebec3e5647491186f8797fde31)
> ##[debug]Download 'https://api.github.com/repos/DoozyX/clang-format-lint-action/tarball/c71d0bf4e21876ebec3e5647491186f8797fde31' to '/home/runner/work/_actions/_temp_f0b37bb0-d8ec-4eaf-b629-48a88581b7fc/a71aa249-f4ce-44d5-8b95-56aacbbe0ba8.tar.gz'
> ##[debug]Unwrap 'DoozyX-clang-format-lint-action-c71d0bf' to '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31'
> ##[debug]Archive '/home/runner/work/_actions/_temp_f0b37bb0-d8ec-4eaf-b629-48a88581b7fc/a71aa249-f4ce-44d5-8b95-56aacbbe0ba8.tar.gz' has been unzipped into '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31'.
> ##[debug]action.yml for action: '/home/runner/work/_actions/actions/checkout/11bd71901bbe5b1630ceea73d27597364c9af683/action.yml'.
> ##[debug]action.yml for action: '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31/action.yml'.
> ##[debug]Dockerfile for action: '/home/runner/work/_actions/DoozyX/clang-format-lint-action/c71d0bf4e21876ebec3e5647491186f8797fde31/Dockerfile'.
> ##[debug]Set step '__actions_checkout' display name to: 'Run actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683'
> ##[debug]Set step '__DoozyX_clang-format-lint-action' display name to: 'Run DoozyX/clang-format-lint-action@c71d0bf4e21876ebec3e5647491186f8797fde31'
> ##[debug]Set step 'diff' display name to: 'Create and uploads code suggestions to apply'
> Complete job name: Formatting Check
> ```
> 
> Compared to the 3 seconds taken to build the dockerfile, and the 15 seconds it takes to run on the repo I use it for (that was way too long with another clang-format action)

That's fast enough for now.  The `~26MB` will keep growing over time. Not sure how fast it'll grow, but it'll slow down over time. I'm fine keeping this branch around until we address the issues with it. Or we can decline this PR for now and just keep on doing things the current way.
